tests/contrib/django/test_middleware.py

```diff
@@ -197,6 +197,8 @@ def test_middleware_trace_error_500(self):
         assert span.get_tag(http.URL) == 'http://testserver/error-500/'
         assert span.resource == 'tests.contrib.django.app.views.error_500'
         assert 'Error 500' in span.get_tag('error.stack')
+        # confirm that the context was properly closed out
+        assert span._context._is_finished()
```

Comment:
Does this actually fail if you remove your `_finalise_span(span, request)` from `process_exception`?

Only reason I am asking is because we will not flush a trace to the writer unless it is fully finished, so I would expect `len(self.tracer.writer.pop()) == 0` if we weren't properly closing the span.

Reply:
https://circleci.com/gh/DataDog/dd-trace-py/125122 This test run was when I added just the assert (and not the fix), you can see the failure here.

To be honest the fact that the test claims that the writer got the trace makes me wonder if the test environment is properly emulating reality. 

When I was testing in a production environment, basically the code ends up calling `Tracer.record`:
```
    def record(self, context):
        """
        Record the given ``Context`` if it's finished.
        """
        # extract and enqueue the trace if it's sampled
        trace, sampled = context.get()
        if trace and sampled:
            self.write(trace)
```
and if you look inside `Context.get`:
```
    def get(self):
        """
        Returns a tuple containing the trace list generated in the current context and
        if the context is sampled or not. It returns (None, None) if the ``Context`` is
        not finished. If a trace is returned, the ``Context`` will be reset so that it
        can be re-used immediately.

        This operation is thread-safe.
        """
        with self._lock:
            if self._is_finished():
                [...]
                return trace, sampled
            elif self._partial_flush_enabled and self._finished_spans >= self._partial_flush_min_spans:
               [...]
                return closed_spans, sampled
            else:
                return None, None
```

So if the tracer is _not_ finished but partial flushing is enabled then things will still work even without this. But if you don't have partial flushing then the context doesn't get recorded to.

So it could be that these tests have partial flushing enabled

Reply:
Though I just said all of this, the test is not passing even with the middleware change... not really sure what's going on here and am having a hard time debugging this.